bblfsh/pyuast.cc

```diff
@@ -1,272 +1,760 @@
+    0,                              // tp_reserved
+    0,                              // tp_repr
+    0,                              // tp_as_number
+    0,                              // tp_as_sequence
+    0,                              // tp_as_mapping
+    0,                              // tp_hash
+    0,                              // tp_call
+    0,                              // tp_str
+    0,                              // tp_getattro
+    0,                              // tp_setattro
+    0,                              // tp_as_buffer
+    Py_TPFLAGS_DEFAULT,             // tp_flags
+    "External UastIterator object", // tp_doc
+    0,                              // tp_traverse
+    0,                              // tp_clear
+    0,                              // tp_richcompare
+    0,                              // tp_weaklistoffset
+    PyUastIterExt_iter,                // tp_iter: __iter()__ method
+    (iternextfunc)PyUastIterExt_next,  // tp_iternext: next() method
+    0,                              // tp_methods
+    0,                              // tp_members
+    0,                              // tp_getset
+    0,                              // tp_base
+    0,                              // tp_dict
+    0,                              // tp_descr_get
+    0,                              // tp_descr_set
+    0,                              // tp_dictoffset
+    0,                              // tp_init
+    PyType_GenericAlloc,            // tp_alloc
+    0,                              // tp_new
+  };
+}
 
-  if (!PyMapping_Check(properties)) {
-    Py_DECREF(properties);
-    return NULL;
-  }
+// ==========================================
+// External UAST Context (managed by libuast)
+// ==========================================
+
+class ContextExt {
+private:
+    uast::Context<NodeHandle> *ctx;
+
+    // toPy allocates a new PyNodeExt with a specified handle.
+    // Returns a new reference.
+    PyObject* toPy(NodeHandle node) {
+        if (node == 0) Py_RETURN_NONE;
+
+        PyNodeExt *pyObj = PyObject_New(PyNodeExt, &PyNodeExtType);
+        if (!pyObj) return nullptr;
+
+        pyObj->ctx = this;
+        pyObj->handle = node;
+        return (PyObject*)pyObj;
+    }
+
+    // toHandle casts an object to PyNodeExt and returns its handle.
+    // Borrows the reference.
+    NodeHandle toHandle(PyObject* obj) {
+        if (!obj || obj == Py_None) return 0;
+
+        if (!PyObject_TypeCheck(obj, &PyNodeExtType)) {
+            const char* err = "unknown node type";
+            PyErr_SetString(PyExc_NotImplementedError, err);
+            ctx->SetError(err);
+            return 0;
+        }
+
+        auto node = (PyNodeExt*)obj;
+        return node->handle;
+    }
+
+    PyObject* newIter(uast::Iterator<NodeHandle> *it, bool freeCtx){
+        PyUastIterExt *pyIt = PyObject_New(PyUastIterExt, &PyUastIterExtType);
+        if (!pyIt)
+          return nullptr;
```

Comment:
According to [C++ styleguide #Conditionals](https://google.github.io/styleguide/cppguide.html#Conditionals)

> In general, curly braces are not required for single-line statements, but they are allowed if you like them;

So it would be nice to be consistent and either have this (and all other cases like this in this file) as one-liner (as above), or add braces [as below](https://github.com/bblfsh/client-python/pull/128/files#diff-8bd44ca8121fa182cb0123e19483c6f4R1068).

Reply:
Fixed in next commit. I've moved to use the one-liner version that was already used sparingly since that's less verbose and also avoids errors accidentally trying to add more code to an unbraced if, except in cases where the line length would be too high.